torch/_refs/fft.py

```diff
@@ -0,0 +1,224 @@
+import torch
+import torch._prims as prims
+import torch._refs as refs
+import torch._prims.utils as utils
+from torch._prims.utils import (
+    TensorLike,
+    TensorLikeType,
+)
+from torch._prims.wrappers import (
+    out_wrapper,
+)
+
+from typing import Union, Tuple, Optional
+from typing_extensions import Literal
+import math
+
+__all__ = ["fft", "ifft", "rfft", "irfft", "hfft", "ihfft"]
+
+NormType = Union[None, Literal["forward"], Literal["backward"], Literal["ortho"]]
+_NORM_VALUES = {None, "forward", "backward", "ortho"}
+
+
+def _apply_norm(
+    x: TensorLike, norm: NormType, signal_numel: int, forward: bool
+) -> TensorLikeType:
+    if norm not in _NORM_VALUES:
+        raise RuntimeError(f"Invalid normalization mode: {norm}")
+
+    if norm == "ortho":
+        return prims.mul(x, 1 / math.sqrt(signal_numel))
+
+    normalize = (not forward and (norm is None or norm == "backward")) or (
+        forward and norm == "forward"
+    )
+    return prims.mul(x, 1 / signal_numel) if normalize else x
+
+
+def _promote_type_fft(
+    dtype: torch.dtype, require_complex: bool, device: torch.device
+) -> torch.dtype:
+    if dtype.is_complex:
+        return dtype
+
+    # Promote integral to default float type
+    if not dtype.is_floating_point:
+        dtype = torch.get_default_dtype()
+
+    is_rocm = False  # TODO: How to discern rocm from CUDA?
```

Comment:
I think we can remove this check. We want trace executors to enable primitive, and if a trace executor (like ROCm) doesn't implement a particular primitive (or primitive x dtype combination, as is the case here), then it should thrown an error when that primitive is called

This may seem odd given that we're suggesting other error checks and input validation happen in the reference implementations and not the primitive implementations, but it's a question of semantic errors vs trace executor errors